### ISSUE TITLE
Deconflicting record with different orderings

### ISSUE THREAD

**Issue by serras** [haskell-works/avro#143]
The [schema resolution spec](https://avro.apache.org/docs/current/spec.html#Schema+Resolution) reads:

> if both are records:
> - the ordering of fields may be different: fields are matched by name.
> - ...

However, the [implementation](https://github.com/haskell-works/avro/blob/master/src/Data/Avro/Schema/Deconflict.hs#L77) mandates the reader schema to be "more specified" than the writer schema.

Is this intended? Should I ensure that my schemas always have such bit on?

**Comment by AlexeyRaga** [comment]
@serras I think there are two different "orderings" here.

One thing is ordering fields in the record. The implementation **does** handle it and **does** perform matching by name. Record fields can in reader and in writer schemas can be specified in different orders. There are tests that ensure this behaviour.

Another thing is the `order` value for each field: it has nothing to do with field's order, but rather indicate how values if this field are sorted: https://avro.apache.org/docs/current/spec.html#order

It says that `Avro binary-encoded data can be efficiently ordered without deserializing it to objects.`, but this library never do it, we always deserialise values. And Haskell's laziness allows us to hopefully not do unnecessary work when parts of the payload are not needed.

However, to be honest, I never used ordered fields and I simply don't know how it should work.

**Comment by serras** [comment]
I get the following error, so I guess that's indeed what's happening, the error is in the sorting of the field `postcode`:

```
Left "Can not resolve differing writer and reader schemas:
(Record {name = \"address\", aliases = [], doc = Nothing, order = Just Ascending, fields = [Field {fldName = \"postcode\", fldAliases = [], fldDoc = Nothing, fldOrder = Just Ascending, fldType = String {logicalTypeS = Nothing}, fldDefault = Nothing},Field {fldName = \"country\", fldAliases = [], fldDoc = Nothing, fldOrder = Just Ascending, fldType = String {logicalTypeS = Nothing}, fldDefault = Nothing}]},
 Record {name = \"address\", aliases = [], doc = Nothing, order = Nothing, fields = [Field {fldName = \"postcode\", fldAliases = [], fldDoc = Nothing, fldOrder = Nothing, fldType = String {logicalTypeS = Nothing}, fldDefault = Nothing},Field {fldName = \"country\", fldAliases = [], fldDoc = Nothing, fldOrder = Nothing, fldType = String {logicalTypeS = Nothing}, fldDefault = Nothing}]})"
```

**Comment by serras** [comment]
Still, the schema resolution page does not mention anything about field ordering.

**Comment by AlexeyRaga** [comment]
I _think_ that what may happen here is this We see that Writer ad Reader Schemas declare different fields ordering. We can deconflict these fields because of the name, but if the order doesn't match then associating data with the reader schema will be a lie, a false assumption and may cause problems.

The specification says that `Data items may only be compared if they have identical schemas.`

I did not implement this functionality, but _i think_ the intuition here is that it is only safe to use reader schema if the reader says "I don't care about the order" or if the order is the same.

I don't know if it is relevant or not for the scope of the library since I have never used `order` myself. Maybe because we currently don't compare anything we should just ignore `order`?

@dwincort  do you have knowledge about this feature?

**Comment by dwincort** [comment]
@AlexeyRaga What you think is happening is exactly what is happening.  In retrospect, I believe it probably shouldn't happen, and we could easily fix the code to allow different orderings by removing `moreSpecified`.

If you recall, when I first redesigned deconflicting, I took a very conservative approach.  In that mindset, it felt strange to accept a reader schema that had a different sort order than the writer schema even though sort order doesn't have anything to do with parsing a particular value.  Your approach of putting some of the coercions directly into `to`/`fromAvro` was much more permissive, as I pointed out in https://github.com/haskell-works/avro/pull/128#issuecomment-572631273, and it made me reconsider just how lenient we can be.  I've definitely come around to the more permissive approach, and I think `moreSpecified` is really just a relic of the earlier design.

**Comment by serras** [comment]
It would be great if you could make this change, indeed, since otherwise the library is not usable for us. Another option is take `Nothig` in `fldOrder` as "don't care", and make the test succeed if any of the parts have a `Nothing`, but not if both are specified and different?

**Comment by AlexeyRaga** [comment]
@dwincort @serras In this case of an agreement I'll remove `moreSpecified` check and will just ignore field's ordering when de-conflicting so that we follow `the ordering of fields may be different: fields are matched by name` rule.

**Comment by AlexeyRaga** [comment]
I think now that we overplay ourselves a bit here :)
We have `Record` defined as:

```haskell
      | Record { name    :: TypeName
               , aliases :: [TypeName]
               , doc     :: Maybe Text
               , order   :: Maybe Order
               , fields  :: [Field]
               }
```

But [Specification for Record](https://avro.apache.org/docs/current/spec.html#schema_record) doesn't mention `order` at the `Record` level at all. `order` should only be at the field level.

I'll remove it and `moreSpecified` will go with it because it uses _that_ field.

**Comment by AlexeyRaga** [comment]
Fixed in #144
